From: =?UTF-8?q?Mika=20Pfl=C3=BCger?= <debian@mikapflueger.de>
Date: Sun, 4 Mar 2012 02:23:13 +0100
Subject: courier policy adjustments: * Label courier socket files as
 courier_var_run_t * Run /usr/sbin/authdaemond as
 courier_authdaemon_t * allow the Courier POP server full
 rw_file_perms access to courier_var_lib_t * Allow courier_pop_t to
 do ioctl on it's fifos. Also allow it to talk to portmap so the
 IMAP server can do FAM * Labelled /var/cache/sqwebmail and allowed
 courier_sqwebmail_t to access it.

---
 policy/modules/services/courier.fc |    4 +++-
 policy/modules/services/courier.if |    2 +-
 policy/modules/services/courier.te |   19 +++++++++++++++----
 3 files changed, 19 insertions(+), 6 deletions(-)

diff --git a/policy/modules/services/courier.fc b/policy/modules/services/courier.fc
index 01d31f1..fceb3a3 100644
--- a/policy/modules/services/courier.fc
+++ b/policy/modules/services/courier.fc
@@ -7,6 +7,7 @@
 /usr/sbin/couriertcpd			--	gen_context(system_u:object_r:courier_tcpd_exec_t,s0)
 
 /usr/lib(64)?/courier/(courier-)?authlib/.*	--	gen_context(system_u:object_r:courier_authdaemon_exec_t,s0)
+/usr/sbin/authdaemond			--	gen_context(system_u:object_r:courier_authdaemon_exec_t,s0)
 /usr/lib(64)?/courier/courier/.*	--	gen_context(system_u:object_r:courier_exec_t,s0)
 /usr/lib(64)?/courier/courier/courierpop.* --	gen_context(system_u:object_r:courier_pop_exec_t,s0)
 /usr/lib(64)?/courier/courier/imaplogin --	gen_context(system_u:object_r:courier_pop_exec_t,s0)
@@ -14,7 +15,8 @@
 /usr/lib(64)?/courier/imapd		--	gen_context(system_u:object_r:courier_pop_exec_t,s0)
 /usr/lib(64)?/courier/pop3d		--	gen_context(system_u:object_r:courier_pop_exec_t,s0)
 /usr/lib(64)?/courier/rootcerts(/.*)?		gen_context(system_u:object_r:courier_etc_t,s0)
-/usr/lib(64)?/courier/sqwebmail/cleancache\.pl -- gen_context(system_u:object_r:sqwebmail_cron_exec_t,s0)
+/usr/lib(64)?/courier/sqwebmail/cleancache\.pl -- gen_context(system_u:object_r:courier_sqwebmail_exec_t,s0)
+/var/cache/sqwebmail(/.*)?			gen_context(system_u:object_r:courier_sqwebmail_cache_t,s0)
 
 ifdef(`distro_gentoo',`
 /usr/lib(64)?/courier-imap/couriertcpd	--	gen_context(system_u:object_r:courier_tcpd_exec_t,s0)
diff --git a/policy/modules/services/courier.if b/policy/modules/services/courier.if
index be99138..351cdaf 100644
--- a/policy/modules/services/courier.if
+++ b/policy/modules/services/courier.if
@@ -29,7 +29,7 @@ template(`courier_domain_template',`
 	allow courier_$1_t self:capability dac_override;
 	dontaudit courier_$1_t self:capability sys_tty_config;
 	allow courier_$1_t self:process { setpgid signal_perms };
-	allow courier_$1_t self:fifo_file { read write getattr };
+	allow courier_$1_t self:fifo_file rw_fifo_file_perms;
 	allow courier_$1_t self:tcp_socket create_stream_socket_perms;
 	allow courier_$1_t self:udp_socket create_socket_perms;
 
diff --git a/policy/modules/services/courier.te b/policy/modules/services/courier.te
index 77eadff..db70f97 100644
--- a/policy/modules/services/courier.te
+++ b/policy/modules/services/courier.te
@@ -24,12 +24,20 @@ files_type(courier_var_lib_t)
 
 type courier_var_run_t;
 files_pid_file(courier_var_run_t)
+files_pid_filetrans(courier_authdaemon_t, courier_var_run_t, { file sock_file })
 
 type courier_exec_t;
 mta_agent_executable(courier_exec_t)
 
+type courier_sqwebmail_cache_t;
+files_type(courier_sqwebmail_cache_t)
+
 courier_domain_template(sqwebmail)
 typealias courier_sqwebmail_exec_t alias sqwebmail_cron_exec_t;
+files_pid_filetrans(courier_sqwebmail_t, courier_var_run_t, { file sock_file })
+
+manage_files_pattern(courier_sqwebmail_t, courier_sqwebmail_cache_t, courier_sqwebmail_cache_t)
+
 dev_read_urand(courier_sqwebmail_t)
 
 ########################################
@@ -46,12 +54,9 @@ allow courier_authdaemon_t courier_tcpd_t:fd use;
 allow courier_authdaemon_t courier_tcpd_t:tcp_socket rw_stream_socket_perms;
 allow courier_authdaemon_t courier_tcpd_t:fifo_file rw_fifo_file_perms;
 
-allow courier_authdaemon_t courier_tcpd_t:tcp_socket rw_stream_socket_perms;
 allow courier_authdaemon_t courier_tcpd_t:unix_stream_socket rw_stream_socket_perms;
 allow courier_authdaemon_t courier_tcpd_t:process sigchld;
 allow courier_authdaemon_t courier_tcpd_t:fd use;
-allow courier_authdaemon_t courier_tcpd_t:tcp_socket rw_stream_socket_perms;
-allow courier_authdaemon_t courier_tcpd_t:fifo_file rw_file_perms;
 
 create_dirs_pattern(courier_authdaemon_t, courier_var_lib_t, courier_var_lib_t)
 manage_sock_files_pattern(courier_authdaemon_t, courier_spool_t, courier_spool_t)
@@ -90,12 +95,18 @@ dev_read_rand(courier_pcp_t)
 # POP3/IMAP local policy
 #
 
-allow courier_pop_t courier_authdaemon_t:tcp_socket rw_stream_socket_perms;
+allow courier_pop_t self:capability { setgid setuid }; 
+allow courier_pop_t courier_authdaemon_t:{ unix_stream_socket tcp_socket } { connectto rw_stream_socket_perms };
 allow courier_pop_t courier_authdaemon_t:process sigchld;
 
 allow courier_pop_t courier_tcpd_t:{ unix_stream_socket tcp_socket } rw_stream_socket_perms;
 dev_read_urand(courier_pop_t)
 
+# for FAM with IMAP
+sysnet_use_portmap(courier_pop_t)
+corenet_tcp_bind_all_rpc_ports(courier_pop_t)
+corenet_tcp_bind_all_nodes(courier_pop_t)
+
 # inherits file handle - should it?
 allow courier_pop_t courier_var_lib_t:file { read write };
 
